### ISSUE TITLE
TypeError('coercing to Unicode: need string or buffer, PosixPath found'

### ISSUE THREAD

**Issue by nipunsadvilkar** [explosion/spaCy#3546]
## How to reproduce the behaviour

With spaCy 2.0 training was happening alright with `json` file-format. 

spacy train command I was using with v2.0:
```
python -m spacy train en anno_model/ train.json test.json -n 10 --no-tagger --no-parser
``` 

> training was done in json format since there was no `jsonl` support in 2.0

After upgrading to newer version v2.1.3 and using `jsonl` format I'm facing following issue.
> I checked source code of `GoldCorpus` but unable to resolve it.

## Your Environment

* Operating System: Ubuntu Linux x86_64
* Python Version Used: 2.7
* spaCy Version Used: v2.1.3

```
(spacy_env) ubuntu@ip-10-8-0-31:~/anno_ner_data$ python -m spacy train en /home/ubuntu/encrypt_volume/anno_spacy/anno_model/ train.jsonl test.jsonl -p ner -n 5 -VV
dropout_from = 0.2 by default
dropout_to = 0.2 by default
dropout_decay = 0.0 by default
batch_from = 100.0 by default
batch_to = 1000.0 by default
batch_compound = 1.001 by default
Training pipeline: [u'ner']
Starting with blank model 'en'
beam_width = 1 by default
beam_density = 0.0 by default
beam_update_prob = 1.0 by default
Counting training words (limit=0)
Exception TypeError: TypeError('coercing to Unicode: need string or buffer, PosixPath found',) in <bound method GoldCorpus.__del__ of <spacy.gold.GoldCorpus object at 0x7f062b4ad750>> ignored
Traceback (most recent call last):
  File "/home/ubuntu/anaconda2/envs/spacy_env/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/home/ubuntu/anaconda2/envs/spacy_env/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/ubuntu/anaconda2/envs/spacy_env/lib/python2.7/site-packages/spacy/__main__.py", line 35, in <module>
    plac.call(commands[command], sys.argv[1:])
  File "/home/ubuntu/anaconda2/envs/spacy_env/lib/python2.7/site-packages/plac_core.py", line 328, in call
    cmd, result = parser.consume(arglist)
  File "/home/ubuntu/anaconda2/envs/spacy_env/lib/python2.7/site-packages/plac_core.py", line 207, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "/home/ubuntu/anaconda2/envs/spacy_env/lib/python2.7/site-packages/spacy/cli/train.py", line 196, in train
    corpus = GoldCorpus(train_path, dev_path, limit=n_examples)
  File "gold.pyx", line 112, in spacy.gold.GoldCorpus.__init__
  File "gold.pyx", line 125, in spacy.gold.GoldCorpus.write_msgpack
KeyError: 1
```

**Comment by ines** [comment]
Thanks for the report. It looks like the `TypeError` isn't the main problem here and is ignored - but we should still look into that.

The `KeyError` seems to be the same as the one reported in #3523, so I'll merge this with the other issue. In the meantime, you could try running `debug-data` and see if it reports anything that looks suspicious.

**Comment by nipunsadvilkar** [comment]
@ines: Error persists for `spacy 2.1.3`.

```
Exception TypeError: TypeError('coercing to Unicode: need string or buffer, PosixPath found',) in <bound method GoldCorpus.__del__ of <spacy.gold.GoldCorpus object at 0x7f24e1888d10>> ignored
```

I used JSON instead of JSONL for training new NER model yet this exception gets raised during saving of the `model-final`/`model-best`.

For a quick fix, I converted relevant `PosixPath` paths to `str` but still seeing the exception. I checked above PR you mentioned but doesn't seem to work out.

Any other fix you suggest?
